openverse_catalog/dags/providers/provider_workflow_types.py

```diff
@@ -0,0 +1,167 @@
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Dict, Optional, Sequence
+
+
+@dataclass
+class ProviderWorkflow:
+    """
+    Required Arguments:
+
+    dag_id:           string giving a unique id of the DAG to be created.
+    provider_script:  string path for the provider_script file whose main
+                      function is to be run. If the optional argument
+                      `dated` is True, then the function must take a
+                      single parameter (date) which will be a string of
+                      the form 'YYYY-MM-DD'. Otherwise, the function
+                      should take no arguments.
+
+    Optional Arguments:
+
+    default_args:      dictionary which is passed to the airflow.dag.DAG
+                       __init__ method.
+    start_date:        datetime.datetime giving the first valid execution
+                       date of the DAG.
+    max_active_tasks:  integer that sets the number of tasks which can
+                       run simultaneously for this DAG, and the number of
+                       dagruns of this DAG which can be run in parallel.
+                       It's important to keep the rate limits of the
+                       Provider API in mind when setting this parameter.
+    schedule_string:   string giving the schedule on which the DAG should
+                       be run.  Passed to the airflow.dag.DAG __init__
+                       method.
+    dated:             boolean giving whether the `main_function` takes a
+                       string parameter giving a date (i.e., the date for
+                       which data should be ingested).
+    day_shift:         integer giving the number of days before the
+                       current execution date the `main_function` should
+                       be run (if `dated=True`).
+    execution_timeout: datetime.timedelta giving the amount of time a given data
+                       pull may take.
+    doc_md:            string which should be used for the DAG's documentation markdown
+    media_types:       list describing the media type(s) that this provider handles
+                       (e.g. `["audio"]`, `["image", "audio"]`, etc.)
+    """
+
+    dag_id: str
+    provider_script: str
+    default_args: Optional[Dict] = None
```

Comment:
Perhaps now might be a good time to tackle this - we either have a binary of "override all arguments, repeating the defaults where necessary" or "just use the defaults". Maybe instead of `dag_defaults = dag_defaults or DAG_DEFAULTS` in the factory function, we could do `dag_defaults = (dag_defaults or {}) | DAG_DEFAULTS` so that we would only have to override the ones we care about? I might be interpreting it all wrong though

Reply:
Ah yes, I didn't notice those factory methods were completely overriding all arguments. It doesn't look like we're actually overriding anything in any of the DAG definitions yet, but that will definitely be a pain when we need to! Done 👍

Reply:
Oh actually wait, I fell into the mutable default trap 😅 _Not_ done, I'll push an update